### DISCUSSION TITLE
running picoli AutoComplete at build time

### DISCUSSION THREAD

**Discussion by apat59** [jbangdev/jbang#863] [Ideas]
For a given script, why isn't jbang doing the picocli AutoComplete call at build time any time it can be done?

Take the cli template and replace the greeting annotation with this:
```
@CommandLine.Option(names = {"-n", "--name"}, description = "Who will we greet?", defaultValue = "World")
private String greeting;
```

Using myScript -h works perfectly, no problem. I'm getting the help page.

But when I run myScript <TAB><TAB> I'd like to see my command line options (here -n --name)

In order to do this, I have to:

```
java -cp "/XXX/.m2/repository/info/picocli/picocli/4.6.1/picocli-.6.1.jar:/XXX/.jbang/cache/jars/myScript.java.972cddd3400c2447a696a2fa08b9962ef4cbf1978491468e5a0f89dfc6980c29.jar"
picocli.AutoComplete -n myScript myScript"
```
then
`source myScript_completion (generated by the previous command)`

and then I'm happy.

Maybe I'm missing something obvious but seems to me that would just be
awesome to have this handled by jbang itself.

**Comment by maxandersen** [comment]
You mean when you do 'jbang app install cuz.java' we setup auto completion into the shell environment?

We would need some kind of convention to know when to do that as Jbang technically don't know your jar is a picocli utility.

I guess Iff you have picocli in your dependency set we could run that autocomplete to generate the completion script and add as part of jbang shell setup. 

Is that What you were thinking about ?

**Reply by apat59** [reply]
The issue is that it takes only few seconds to have an executable 'java' source with jbang and it's 200% user friendly.
However enabling completion is tricky. You must first generate the cuz_completion file and source it.

So it can be many things:
- doc improvement, and explain how one can generate it
- + a log the 1st time it builds
- always generate that cuz_completion file and explain what it is meant for. If jbang doesn't generate it, that means we expect the user to dig into picocli documentation, deal with jars, ...
- 'jbang app install cuz.java' would be really nice indeed

In any case, users must quickly get the answer to "how do I enable completion for my scripts if I use any cli template?"

**Comment by quintesse** [comment]
This is somewhat similar to this issue #788
Except that it's for installed commands which makes it more complicated (for #788 we know that Jbang has completions but like mentioned by @maxandersen we don't know that for third party scripts. It also means potentially having to change the user's environment on every install)

**Reply by maxandersen** [reply]
if we do it what I would do is have our initial script just find all ~/.jbang/bin/*.completions and source them.

those .completions would show up when you do something like `jbang app install --completions myapp.java`

then yes, completion might be slightly out of date BUT it would be very fast.

**Reply by odrotbohm** [reply]
For installations from (Maven) artifacts, would it make sense to check if the JAR contained a `.completion` file in a well-known location (e.g., `META-INF/jbang/*.completion`) within the JAR itself and apply that?

**Reply by maxandersen** [reply]
Yes - as long as have a way to handle the various shell types. 

though here five years later im inclined to look at https://github.com/jdx/usage as a mechanism so it's not tied to only picocli driven clis. 

But yeah. If some can come up with something not too hairy to make this work be my guest.
